### PR TITLE
Generate templates with an option to preserve existing changes

### PR DESCRIPTION
**Pull request by ghost**
changes to `config:generate_all` so it allows user the option to review the diff and preserve any project specific customization.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -109,9 +109,18 @@ directory are provided:
 Run the following command to generate them:
 
 ```
-[bundle exec] rake config:generate_all
+[bundle exec] rake config:generate_all
```
Comment (jcarres-mdsol) [inline README.md]: extra space

File: README.md [hunk 2]
```diff
@@ -109,9 +109,18 @@ directory are provided:
...
-[bundle exec] rake config:generate_all
+[bundle exec] rake config:generate_all  
 ```
 
+This command provides options to compare changes between source and local templates, so new additions to the source templates can be safely added while preserving any project specific customization to local templates.
+
+Alternatively, to force generate templates (replacing existing local templates with the source), run the following: 
+
+```
+[bundle exec] rake config:generate_all[true]
```
Comment (jcarres-mdsol) [inline README.md]: :+1:  very cool idea

File: dice_bag.gemspec [hunk 3]
```diff
@@ -15,6 +15,8 @@ Gem::Specification.new do |s|
   s.require_paths = ['lib']
 
   s.add_dependency 'rake'
+  s.add_dependency 'thor', '~> 0'
```
Comment (jcarres-mdsol) [inline dice_bag.gemspec]: maybe ~> 0.0 is better because if version 1.0 is ever released may have breaking changes

File: dice_bag.gemspec [hunk 4]
```diff
@@ -15,6 +15,8 @@ Gem::Specification.new do |s|
...
+  s.add_dependency 'diff-lcs'
```
Comment (jcarres-mdsol) [inline dice_bag.gemspec]: shoudl have some restriction

File: lib/dice_bag/tasks/config.rake [hunk 5]
```diff
@@ -23,15 +23,15 @@ namespace :config do
   end
 
   desc "Generate all configuration file templates needed by this project"
-  task :generate_all do
-    DiceBag::Command.new.generate_all_templates
+  task :generate_all, :force do |t, args|
```
Comment (jcarres-mdsol) [inline lib/dice_bag/tasks/config.rake]: does this make :force a required param?

File: lib/dice_bag/tasks/config.rake [hunk 6]
```diff
@@ -23,15 +23,15 @@ namespace :config do
...
   end
 
   desc "Generate just the specified template, optionally in the specified location"
-  task :generate, :filename, :location do |t, args|
+  task :generate, :filename, :location, :force do |t, args|
     filename = args[:filename]
     location = args[:location]
     raise "A filename needs to be provided" if filename.nil?
-    DiceBag::Command.new.generate_template(DiceBag::DefaultTemplateFile.new(filename,location))
+    DiceBag::Command.new.generate_template(DiceBag::DefaultTemplateFile.new(filename, location), args[:force])
```
Comment (jcarres-mdsol) [inline lib/dice_bag/tasks/config.rake]: I'm guessing if not provided args[:force] would be nil . Maybe you should do  !!args[:force] to make that nil become 'false' instead

File: lib/dice_bag/command.rb [hunk 7]
```diff
@@ -25,15 +32,14 @@ def write(template_name, params = {})
...
-        generate_template(template)
+        generate_template(template, force)
       end
     end
 
-    def generate_template(default_template)
-      default_template.assert_existence
-      default_template.create_file
+    def generate_template(default_template, force=false)
+      copy_file default_template.file, default_template.destination, force: force
```
Comment (jcarres-mdsol) [inline lib/dice_bag/command.rb:42]: does copy_file puts some message when it just copies over and there was no file to overwrite?
I just want to have some notification that things indeed worked.

File: lib/dice_bag/templates/database.yml.dice [hunk 8]
```diff
@@ -7,10 +7,10 @@
...
   password: <%= configured[env].database_password %>
   host: <%= configured[env].database_host || 'localhost' %>
-  pool: 5
-  timeout: 5000
-  encoding: utf8
-  reconnect: false
+  pool: <%= configured[env].database_pool || 5 %>
+  timeout: <%= configured[env].database_timeout || 5000 %>
+  encoding:  <%= configured[env].database_encoding || 'utf8' %>
+  reconnect:  <%= configured[env].database_reconnect || false %>
```
Comment (jcarres-mdsol) [inline lib/dice_bag/templates/database.yml.dice:13]: :+1:  good changes

File: lib/dice_bag/command.rb (hunk 7 above)
Comment (ghost) [inline lib/dice_bag/command.rb:42]: it prints the action performed on specific files, e.g.
if new file created: 

``` bash
create  config/dalli.yml.dice
```

if modified: 

``` bash
force  config/dalli.yml.dice
```

if no changes: 

``` bash
identical  config/dalli.yml.dice
```

**Comment by ghost** [comment]
@jcarres-mdsol im thinking maybe its cleaner if I create separate 'force' generate tasks 

``` ruby
config:generate_all:force
config:generate:force[filename, location]
```

This will get rid of the additional boolean input. any thoughts?

File: lib/dice_bag/command.rb (hunk 7 above)
Comment (jcarres-mdsol) [inline lib/dice_bag/command.rb:42]: very cool

**Comment by jcarres-mdsol** [comment]
@atierney-mdsol on one hand sounds good, easier to call. On the other hand we already have a bunch of tasks, I think it will be confusing for users to have too many tasks to choose from and pollute the rails tasks.
Looking at the tasks file in general what's easier for you as user?

**Comment by atierney-mdsol** [comment]
@atrivedi-mdsol think I got tagged in by mistake :smile:

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,3 +1,8 @@
+# 0.10.0
```
Comment (jcarres-mdsol) [inline CHANGELOG.md]: I think you can pretty much declare 1.0.0 , as we for all purposes we are treating this gem as semver, no point of staying in 0.x versions

**Comment by ghost** [comment]
@jcarres-mdsol I did a very extensive poll (i.e. asked 2 people at work) about which task seem easier. they seem to prefer `config:generate_all:force` over `config:generate_all[true]`. I agree, personally i think the `:force` variant reads better and is more explicit about what it does. it also removes the extra code we need to put in to verify user input.

**Comment by jcarres-mdsol** [comment]
@atrivedi-mdsol  The poll seems extensive enough to me.
Your PR is perfect.
 Travis seems to have a weird problem though. Maybe is a hiccup, can you do one more commit to make it run tests again?
Maybe you can add ruby 2.2 to the travis.yml file. We need that anyways. Also .ruby_version should probably be ruby 2.0 at least.

**Comment by ghost** [comment]
Thanks @jcarres-mdsol. I updated travis.yml and .ruby-version 
travis isn't playing nice with ruby 1.9.3 build. any idea what causes this error? (works locally with 1.9.3)

``` ruby
NoMethodError: undefined method `spec' for nil:NilClass
An error occurred while installing dice_bag (1.0.0), and Bundler cannot
continue
```

Some gem dependency you reckon?

**Comment by ghost** [comment]
looks like CI failure is caused by bundler/bundler#3559

**Comment by jcarres-mdsol** [comment]
Probably can be solved by updating to latest bundler version before running
the tests.
Something like adding this to the .travis.yml

install:
- gem uninstall bundler
- gem install bundler --version '1.11'

On Wed, Dec 16, 2015 at 9:48 PM, Abhi Trivedi notifications@github.com
wrote:

> looks like CI failure is caused by bundler/bundler#3559
> https://github.com/bundler/bundler/pull/3559
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/mdsol/dice_bag/pull/75#issuecomment-165234319.

## 

_Jordi Polo Carres_ | Software Architect  | Medidata Solutions Worldwide
http://www.mdsol.com/

**Comment by jcarres-mdsol** [comment]
Oh, it seems it passed. I'm having problems with Travis and bundler this couple of days, maybe they solved them. Merging! :tada:

**Pull request merged**